File: resources/lorapayload/payloads/bmeters_iwmlr3.py
# This file is part of Jeedom.
#
# Jeedom is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Jeedom is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Jeedom. If not, see <http://www.gnu.org/licenses/>.
#
import logging
import struct
import globals
import time
import json
import threading
import utils

class bmeters_iwmlr3():
	def __init__(self):
		self.name = 'bmeters_iwmlr3'

	def parse(self,data,device):
		logging.debug('IWM LR3 Received')
		parsed={}
		data['parsed'] = parsed
		try:
			logging.debug('Parsing')
			payload = data['payload']

			def convertHex(value):
			   little_hex = bytearray.fromhex(value)
			   little_hex.reverse()
			   hex_string = ''.join(format(x, '02x') for x in little_hex)
			   #logging.debug(hex_string)
			   return hex_string

			appCode = payload[0:2]
			if appCode == '44':
				parsed['absValue'] = convertHex(payload[2:10])
				parsed['revFlow'] = convertHex(payload[10:18])
				indexK_value = payload[18:20]
				if indexK_value == '00':
					K = 1
				elif indexK_value == '01':
					K = 10
				elif indexK_value == '02':
					K = 100
				
				medium_value = payload[20:22]
				if medium_value == '00':
					parsed['mediumValue'] = 'Water'
				elif medium_value == '01':
					parsed['mediumValue'] = 'Hot Water'

				parsed['vif'] = payload[22:24] * K

				alarm_byte = int(payload[24:26], 16)
				if alarm_byte & 0x01:
					parsed['alarmMagnetic'] = 1
				else:
					parsed['alarmMagnetic'] = 0

				if alarm_byte & 0x02:
					parsed['alarmRemoval'] = 1
				else:
					parsed['alarmRemoval'] = 0

				if alarm_byte & 0x04:
					parsed['alarmSensorFraud'] = 1
				else:
					parsed['alarmSensorFraud'] = 0

				if alarm_byte & 0x08:
					parsed['alarmLeakage'] = 1
				else:
					parsed['alarmLeakage'] = 0

				if alarm_byte & 0x10:
					parsed['alarmReverseFlow'] = 1
				else:
					parsed['alarmReverseFlow'] = 0

				if alarm_byte & 0x20:
					parsed['alarmLowBattery'] = 1
				else:
					parsed['alarmLowBattery'] = 0

				temp_raw = int(payload[26:30], 16)
				if temp_raw & 0x8000:
					temp_value = -(temp_raw & 0x7FFF) / 10.0
				else:
					temp_value = temp_raw / 10.0
				parsed['temp'] = temp_value

			elif appCode == '07':
				logging.debug('Get firmware version')
				parsed['firmware'] = int(payload[-6:], 16)

			elif appCode == '17':
				logging.debug('GET_REVOLUTION_COUNTERS')
				parsed['fwdCnt'] = int(payload[12:20], 16)
				parsed['backwardCnt'] = int(payload[20:22], 16)

		except Exception as e:
			logging.debug(str(e))
			logging.debug('Unparsable data')
		data['parsed'] = parsed
		return data

globals.COMPATIBILITY.append(bmeters_iwmlr3)

File: resources/lorapayload/payloads/bmeters_iwmlr3.js
/* https://github.com/IoTThinks/lorawan-devices/blob/master/vendor/b-meters/iwmlr3.js */
function Decode(input) {
  input.fPort = 1;
  var data = {};

  function toHexLE(bytes) {
      return bytes
          .slice()
          .reverse()
          .map(b => b.toString(16).padStart(2, '0'))
          .join('');
  }

  function bytesToIntBE(bytes) {
      return bytes.reduce((acc, b) => (acc << 8) + b, 0);
  }

  function bcdToInt (b) {
    return ((b >> 4) & 0x0f) * 10 + (b & 0x0f);
  }

  switch (parseInt(input.fPort, 10)) {
      case 1:
          var appCodeHex = input.bytes[0].toString(16).padStart(2, '0');
          data.application = appCodeHex;

          if (appCodeHex === '44') {
              data.absValue   = toHexLE(input.bytes.slice(1, 5));
              data.revFlow    = toHexLE(input.bytes.slice(5, 9));
              var idxK = input.bytes[9];
              var K = idxK === 0x01 ? 10 : idxK === 0x02 ? 100 : 1;
              var med = input.bytes[10];
              data.mediumValue = med === 0x00 ? 'Water' : med === 0x01 ? 'Hot Water' : 'Unknown';
              data.vif        = bcdToInt(input.bytes[11]) * K;
              var alarm       = input.bytes[12];
              data.alarmMagnetic    = (alarm & 0x01) ? 1 : 0;
              data.alarmRemoval     = (alarm & 0x02) ? 1 : 0;
              data.alarmSensorFraud = (alarm & 0x04) ? 1 : 0;
              data.alarmLeakage     = (alarm & 0x08) ? 1 : 0;
              data.alarmReverseFlow = (alarm & 0x10) ? 1 : 0;
              data.alarmLowBattery  = (alarm & 0x20) ? 1 : 0;
              var tempRaw     = bytesToIntBE(input.bytes.slice(13, 15));
              var isNeg       = (tempRaw & 0x8000) !== 0;
              data.temp       = (isNeg ? -(tempRaw & 0x7FFF) : tempRaw) / 10.0;
          }
          else if (appCodeHex === '07') {
              var last3 = input.bytes.slice(-3);
              data.firmware = bytesToIntBE(last3);
          }
          else if (appCodeHex === '17') {
            var rawFwd = bytesToIntBE(input.bytes.slice(6, 10));
            var unitBits = rawFwd >>> 30;
            var count = rawFwd & 0x3FFFFFFF;
            var factor = unitBits === 0 ? 1
                      : unitBits === 1 ? 10
                      : unitBits === 2 ? 100
                      : null;
            data.fwdCnt = factor !== null
              ? count * factor
              : 'Invalid unit code';
            data.backwardCnt = input.bytes[10];
          }
          else {
              data.error = 'Unknown appCode 0x' + appCodeHex;
          }

          return { data };

      default:
          return { errors: ['unknown FPort'] };
  }
}

module.exports = {
  Decode
};
